llvm/lib/Transforms/Utils/CodeLayout.cpp

```diff
@@ -798,17 +811,15 @@ class ExtTSPImpl {
   /// element being the corresponding merging type.
   MergeGainT getBestMergeGain(ChainT *ChainPred, ChainT *ChainSucc,
                               ChainEdge *Edge) const {
-    if (Edge->hasCachedMergeGain(ChainPred, ChainSucc)) {
+    if (Edge->hasCachedMergeGain(ChainPred, ChainSucc))
       return Edge->getCachedMergeGain(ChainPred, ChainSucc);
-    }
 
     // Precompute jumps between ChainPred and ChainSucc.
-    auto Jumps = Edge->jumps();
-    ChainEdge *EdgePP = ChainPred->getEdge(ChainPred);
-    if (EdgePP != nullptr) {
-      Jumps.insert(Jumps.end(), EdgePP->jumps().begin(), EdgePP->jumps().end());
-    }
+    MergedVector<JumpT> Jumps(Edge->jumps().begin(), Edge->jumps().end());
     assert(!Jumps.empty() && "trying to merge chains w/o jumps");
+    ChainEdge *EdgePP = ChainPred->getEdge(ChainPred);
+    if (EdgePP != nullptr)
+      Jumps.append(EdgePP->jumps().begin(), EdgePP->jumps().end());
```

Comment:
It's not clear why MergedVector is used for Jumps. I think it's better to define a separate struct for it.

Reply:
I am not sure i understand this suggestion, can you clarify?

Reply:
IIUC, `Jumps` is more restricted than this `MergedVector` type. For instance, it doesn't need `MergedVector::mergeNodes`. The only function it uses is `ForEach`. However, it uses ibe function that the other use case (chains of nodes) does not need (`append`). So why don't we define a separate struct for it?

Reply:
Separated the structs as suggested

Reply:
@rlavaee let me know if you think some other refactoring is needed here. Thanks